Feat 4: Setup safe reality

fabianschu opened the pull request:
closes #4 

**Acceptance Criteria**

- [x] hardhat tasks:
  - [x] deploy & configure Zodiac Reality
  - [x] deploy Shrine
  - [x] configure Reality oracle
  - [x] propose an airdrop via Zodiac Reality
  - [x] answer the Reality oracle (Reality's "economic game")
  - [x] execute a finalized airdrop proposal via Zodiac Reality

- [x] e2e test: deploy contracts =>  configure contracts => propose merkle drop => play the reality.eth "economic game" => execute the merkle drop => claim a drop as a user

**Additional Info**

These parts might be relevant for @codebuster22:
- module to construct a merkle tree: https://github.com/Curve-Labs/ie-bacalhau/blob/feat-4/setup-safe-reality/packages/on-chain/utils/merkleTree.ts
- how the merkle tree is constructed: https://github.com/Curve-Labs/ie-bacalhau/blob/feat-4/setup-safe-reality/packages/on-chain/test/airdrop.behavior.spec.ts#L121-L125
- what should be stored on ipfs: https://github.com/OpenZeppelin/merkle-tree#treedump

Relevant docs:
- https://github.com/gnosis/zodiac-module-reality
- https://realitio.github.io/docs/html/index.html
- https://github.com/Astrodrop/shrine

fabianschu committed:
feat: sets up test avatar and reality module

fabianschu committed:
chore: adds shrine contracts

fabianschu committed:
chore: installs types

fabianschu committed:
chore: replaces outdated realitio

fabianschu committed:
chore: allows import of json (ts)

fabianschu committed:
chore: adds test token

fabianschu committed:
feat: adds setup specs

fabianschu committed:
feat: adds specs for initiating merkle drop via reality

fabianschu committed:
chore: adds package for merkle tree construction (js)

fabianschu committed:
chore: removes redundant hardhat-deploy

fabianschu committed:
feat: tests for claiming airdrop

fabianschu committed:
feat: adds task for answering question on reality

fabianschu committed:
chore: removes faulty ts config

fabianschu committed:
chore: format w/ prettier

fabianschu committed:
chore: adds comments

Zitzak commented on packages/on-chain/tasks/reality.ts:
```diff
@@ -0,0 +1,523 @@
+    types.string
+  )
+  .addParam("amount", "Amount of the token to be offered", undefined, types.int)
+  .setAction(
```
I think this file gets quite large like this. It is not required on my side, seeing the time constraints, but I think it might make sense to add a `lib` directory with a file called `reality.ts` which has the different functions in it. Within that lib directory you could also abstract different things that you repeat in the tasks, like getting caller, the reality module instance etc.

fabianschu replied:
I looked through it and found `reatliy:assemble` repetitive, so I refactored. Can you give me an example of something else that you perceive as repetitive or non-descriptive?

Zitzak replied:
I think what I had in mind was overkill for these tasks. Looks good to me the way it is

Zitzak commented on packages/on-chain/test/airdrop.behavior.spec.ts:
```diff
@@ -0,0 +1,220 @@
+  const user1Amount = "50";
+  const user2Amount = "50";
+
+  async function deployFixture() {
```
Not sure if fixtures are needed in other test files as well, but if so, it might be smart to move it to a separate file. In that way it can be added to and each test file only uses the object members they need from the return values

codebuster22 replied:
+1

fabianschu replied:
This is the only test file.

Zitzak commented:
PR looks good, although I am missing a lot of context on Zodiac/Shrine to dive into it deeply. I added two comments about restructuring/abstracting parts to make it more clear but those are not strictly necessary seeing the time constraints.

One other idea for review ease is to also link some relevant Zodiac documentation from which you based certain tasks, if you have any, so I can have a look at it
@fabianschu

codebuster22 commented on packages/on-chain/contracts/zodiac/RealityModule.sol:
```diff
@@ -0,0 +1,533 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
should we mention that this file was imported from other repo.
maybe something like this
```
File imported/taken directly from ${URL} as it is with no modification.
```
or maybe a better way to put it.

Might help in reviewing as well as reference for others.

codebuster22 commented on packages/on-chain/contracts/interfaces/RealitioV3.sol:
```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
should we mention that this file was imported from other repo.
maybe something like this
```
File imported/taken directly from ${URL} as it is with no modification.
```
or maybe a better way to put it.

Might help in reviewing as well as reference for others.

codebuster22 commented on packages/on-chain/contracts/zodiac/RealityModuleERC20.sol:
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
same as above

codebuster22 commented on packages/on-chain/contracts/zodiac/RealityModuleETH.sol:
```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
same as above

codebuster22 commented on packages/on-chain/hardhat.config.ts:
```diff
@@ -1,8 +1,29 @@
+  external: {
+    contracts: [
+      {
+        artifacts: "../node_modules/@gnosis.pm/zodiac/dist/src/abi",
```
Nice :)

fabianschu replied:
was removed in later commits because it wasn't useful^^

codebuster22 commented on packages/on-chain/test/setup.behavior.spec.ts:
```diff
@@ -0,0 +1,151 @@
+    const cooldown = 23;
+    const expiration = 0;
+    const bond = 0;
+    const templateId = 1337;
```
is this same as chain ID?

fabianschu replied:
no, this is an arbitrary id

codebuster22 commented on packages/on-chain/contracts/astrodrop/Shrine.sol:
```diff
@@ -0,0 +1,654 @@
+// SPDX-License-Identifier: AGPL-3.0
```
Same comment as for Reality contracts.

codebuster22 commented on packages/on-chain/contracts/astrodrop/lib/Ownable.sol:
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: AGPL-3.0
```
same as above

codebuster22 commented on packages/on-chain/contracts/astrodrop/lib/ReentrancyGuard.sol:
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: AGPL-3.0-only
```
same as above

codebuster22 reviewed (dismissed):
currently reviewed 2 commits, will do more in abit.

fabianschu committed:
chore: adds comments on contracts that have been copied

fabianschu committed:
chore: refactor reality:assemble

fabianschu commented:
> PR looks good, although I am missing a lot of context on Zodiac/Shrine to dive into it deeply. I added two comments about restructuring/abstracting parts to make it more clear but those are not strictly necessary seeing the time constraints.
> 
> One other idea for review ease is to also link some relevant Zodiac documentation from which you based certain tasks, if you have any, so I can have a look at it @fabianschu

I added links to relevant documentations to the PR description.

Zitzak reviewed (approved):
LGTM
